upstream-seed/src/peer.rs

```diff
@@ -0,0 +1,163 @@
+// Copyright © 2021 The Radicle Upstream Contributors
+//
+// This file is part of radicle-upstream, distributed under the GPLv3
+// with Radicle Linking Exception. For full terms see the included
+// LICENSE file.
+
+use std::{net::SocketAddr, time::Duration};
+
+use anyhow::Context as _;
+use futures::stream::StreamExt as _;
+
+use librad::{
+    git::{identities, replication, storage::fetcher, tracking},
+    net::{
+        discovery::{self, Discovery as _},
+        protocol, Network,
+    },
+    paths::Paths,
+    PeerId,
+};
+use link_identities::git::Urn;
+
+type LibradPeer = librad::net::peer::Peer<librad::SecretKey>;
+type PeerInfo = librad::net::peer::PeerInfo<SocketAddr>;
+
+/// Configuration for creating a new [`Peer`].
+#[derive(Clone)]
+pub struct Config {
+    pub rad_paths: Paths,
+    pub key: librad::SecretKey,
+    pub listen: SocketAddr,
+}
+const PROVIDER_QUERY_TIMEOUT: Duration = Duration::from_secs(60);
+
+/// Wrapper around [`librad::net::peer::Peer`] that provides seed specific functionality.
+#[derive(Clone)]
+pub struct Peer {
+    librad_peer: LibradPeer,
+}
+
+impl Peer {
+    /// Create a new client.
+    pub fn new(config: Config) -> Self {
+        let storage = librad::net::peer::config::Storage {
+            protocol: librad::net::peer::config::ProtocolStorage {
+                fetch_slot_wait_timeout: Default::default(),
+                pool_size: 4,
```

Comment:
What determines the pool size? Number of VCPUs?

Reply:
I’m not sure, I only copied this from the org node. I believe this controls how many threads can interact with the storage concurrently. This has probably more to do with IO than with CPUs.